Setup lerna

ang-zeyu opened the pull request:
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**

• [x] Other, please explain:

Resolves https://github.com/MarkBind/markbind/issues/1285



**What is the rationale for this request?**
See #1285

**What changes did you make? (Give an overview)**
Commit order:
- Move all cli files to `packages/cli` 5a56a69
  - and update `.gitignore` ignore path accordingly
- Add`.npmignore` for core package d7ca8bc
- Setup lerna (gitignore update + `lerna init` + `lerna bootstrap`) 7979698
- Update font awesome test files (the previous step caused some package-lock versions to change, which caused font awesome to update 605a4e2
- Adapt test scripts f7c16e3
- Adapt linting configs, scripts c3c22bc
  - this commit also unifies the linting into one call to `eslint .`, and ditches separately running `eslint .` in all packages. This allows vue-components' eslint to properly extend the root config
- Move vue-components script to root private package 823e9ec
- Adapt CI and setup (install) scripts b0d33ca
- Update devguide bbe4bec (wip)
  - the publishing process is yet to be tested here

**Is there anything you'd like reviewers to focus on?**
* bbe4bec 's publishing process is yet to be dummy tested

**Testing instructions:**
- no change -- `npm run test`
  - under the hood this now runs eslint & stylelint, then the `test` npm script in all packages separately

**Proposed commit message: (wrap lines at 72 characters)**
The cli and core packages uses relative file specifiers to link
dependencies in the mono repo. This causes a published cli package to
fail to install.

Besides site generation, the core package also houses several other
things like templates, assets and plugins, which can be moved into
separate packages for better separation of concerns.

Let's extract the main cli package into packages/cli as such, resulting
in a flat package structure.
The multi package structure implies publishing all packages
independently, which fixes the usage of relative file specifiers.

Let's also setup lerna, which is a multi-package monorepo development
and dependency management helper.
This greatly reduces the amount of work needed to manage multiple
independent packages.

marvinchin commented on .eslintignore:
```diff
@@ -2,6 +2,33 @@
 
 !.eslintrc.js
+
+# vue-components
```
Just wondering - is there any reason why we don't just use `packages/vue-components` here?

ang-zeyu replied:
👍 fixed

ang-zeyu replied:
> Some questions/thoughts:
> - Do you know why the FontAwesome version was bumped? Ideally, we should retain the original version in this PR to keep these concerns separate and avoid unnecessary diffs.

we previously installed the core package transitively via the cli package, so there was no `package-lock` there. so the latest semantic ver gets installed.

I'll get out a separate pr to update it though, then rebase this one on top of that

> - I feel that we should keep the `README.md` in the root. This can help new devs who first visit our repo some information on the project and how to get started at first glance. Maybe we should set up a new README for `cli` and each of the other leaf packages, with a short explanation on what each package does.

🤦‍♂️ unintentional. Will revert and add some brief readmes for `cli/core` after the `npmignore` commit during the rebase above

marvinchin commented on docs/devGuide/settingUp.md:
```diff
@@ -32,14 +32,16 @@ We recommend the **WebStorm IDE** for working with MarkBind code.
-   <popover header="Why **2** `ci` commands?" content="Our components subpackage depends on `node-sass` indirectly, which depends on `fsevents`.<br><br> `fsevents` is not needed nor compatible with Windows, and `npm ci` does not respect this since it uses `package-lock.json`.">(`ciwin:all` if you are on a Windows machine)</popover>
-   in the root folder of your cloned repo.
-1. **To bind your cloned version of MarkBind to your console** (instead of the released version of MarkBind), run `npm link` in the root folder of the cloned repo.
+1. **Install dependencies** by running <popover content="Under the hood, this calls `npm ci` and `lerna bootstrap`">`npm run setup`</popover> in the root folder of your cloned repo.
```
I had some issues due to existing `node_modules` directories in the existing packages. Do you think it's worth adding a note here for existing markbind devs to first run `lerna clean` before running `npm run setup`?

ang-zeyu replied:
`lerna clean` won't work unless you have it installed globally or have run `npm ci` (which `npm run setup` does) first unfortunately. I think the devguide should be written in the perspective of new developers always though. Experienced developers could always drop an issue / ping, and have a likely have a better understanding of node already 😁

marvinchin replied:
ah fair enough!

marvinchin commented on docs/devGuide/settingUp.md:
```diff
@@ -32,14 +32,16 @@ We recommend the **WebStorm IDE** for working with MarkBind code.
+  MarkBind uses <md>[lerna](https://github.com/lerna/lerna)</md>, a popular multi-package development tool, to manage it's dependencies. It is essentially a high level wrapper over node and npm's functionalities
+  </box>
+  
+1. **To bind your cloned version of MarkBind to your console** (instead of the released version of MarkBind), run `npm link` in the `packages/cli` folder of the cloned repo.
```
Does this work for you now? I'm getting errors because our internal packages haven't been published yet.

ang-zeyu replied:
got that too, you could get around it by temporarily removing the `@markbind/core` dependency and `npm link` then undoing it

ang-zeyu commented:
Updated:
- reverted README.md movement 502041c
- add README for cli / core package 6360743
- dropped font awesome commit

---

https://github.com/lerna/lerna/blob/master/doc/hoist.md

Also switched over to `lerna bootstrap --hoist` in `npm run setup` (`package-lock` changes in 502041c & d6b62c3 only), which is compatible with the current setup and should install faster + take up less space.

the second disadvantage mentioned isn't applicable either as `eslint-plugin-import` guards against it

marvinchin commented on packages/cli/package-lock.json:
```diff
@@ -0,0 +1,4683 @@
+{
```
I presume that these package-lock changes are done by `lerna bootstrap`? Ideally we'd be able to verify that our dependency versions are still the same, but these diffs seem a little too large to manually verify 🙁

ang-zeyu replied:
yep, I think that if our functional tests stay the same we should be good though, since `package-lock` is only effective in development environment

if there are changes (like the font awesome bundles), we can then investigate them and deal with the diff in a separate pr

ang-zeyu commented on packages/vue-components/src/Searchbar.vue:
```diff
@@ -21,9 +21,9 @@ export default {
       }
       const pages = [];
       const regexes = this.value.split(' ')
-        .filter((searchKeyword) => searchKeyword !== '')
```
yup the vue eslintrc now properly extends the previous root config

ang-zeyu commented:
Updated @marvinchin, seems to work. Travis seems to be broken on all prs though, likely unrelated to the changes here

494ba03 - rename the packages with `markbind-` prefix
4dc983d - changed `lerna publish` to -> `lerna publish from-git` ( tells lerna to use the version commit from `lerna version` instead of running `lerna version` again )

marvinchin commented:
This is a big PR with multiple steps. I feel that we should do a merge commit, rather than squashing the commit history here.

I also think we should add some detail to commit 494ba03 explaining why we are doing the renaming.

ang-zeyu commented:
> This is a big PR with multiple steps. I feel that we should do a merge commit, rather than squashing the commit history here.

yup that's the plan

> I also think we should add some detail to commit [494ba03](https://github.com/MarkBind/markbind/commit/494ba03d56ff767a9d32f88d03c06a5f37efb983) explaining why we are doing the renaming.

Just received npm permissions from @yamgent, will revert back to `@markbind/xx`!

ang-zeyu commented:
reverted to `@markbind`, should be ready to merge now. Thanks @marvinchin @yamgent!
